presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java

```diff
@@ -0,0 +1,154 @@
+    private static final MethodHandle METHOD_HANDLE = methodHandle(ArrayIntersectFunction.class, "intersect", Type.class, Slice.class, Slice.class);
+
+    @Override
+    public Signature getSignature()
+    {
+        return SIGNATURE;
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return true;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Intersects elements of the two given arrays.";
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        checkArgument(types.size() == 1, format("%s expects only one argument", FUNCTION_NAME));
+        TypeSignature typeSignature = parameterizedTypeName("array", types.get("E").getTypeSignature());
+        MethodHandle methodHandle = METHOD_HANDLE.bindTo(types.get("E"));
+        Signature signature = new Signature(FUNCTION_NAME, typeSignature, typeSignature, typeSignature);
+        return new FunctionInfo(signature, getDescription(), isHidden(), methodHandle, isDeterministic(), false, ImmutableList.of(false, false));
+    }
+
+    public static Slice intersect(Type type, Slice leftArray, Slice rightArray)
+    {
+        Block leftBlock = readStructuralBlock(leftArray);
+        Block rightBlock = readStructuralBlock(rightArray);
+
+        int leftPositionCount = leftBlock.getPositionCount();
+        int rightPositionCount = rightBlock.getPositionCount();
+        int leftPosition = 0;
+        int rightPosition = 0;
+
+        List<Integer> leftPositions = Ints.asList(new int[leftPositionCount]);
+        List<Integer> rightPositions = Ints.asList(new int[rightPositionCount]);
+
+        for (int i = 0; i < leftPositionCount; i++) {
+            leftPositions.set(i, i);
+        }
+        for (int i = 0; i < rightPositionCount; i++) {
+            rightPositions.set(i, i);
+        }
+
+        Collections.sort(leftPositions, new Comparator<Integer>()
+        {
+            @Override
+            public int compare(Integer l1, Integer l2)
+            {
+                return type.compareTo(leftBlock, l1, leftBlock, l2);
+            }
+        });
+        Collections.sort(rightPositions, new Comparator<Integer>()
+        {
+            @Override
+            public int compare(Integer r1, Integer r2)
+            {
+                return type.compareTo(rightBlock, r1, rightBlock, r2);
+            }
+        });
+
+        BlockBuilder leftBlockBuilder = new VariableWidthBlockBuilder(new BlockBuilderStatus(), leftBlock.getSizeInBytes());
+        BlockBuilder rightBlockBuilder = new VariableWidthBlockBuilder(new BlockBuilderStatus(), rightBlock.getSizeInBytes());
```

Comment:
Building these blocks will be a bit expensive. Instead I would fuse it with the loop below, by iterating over the leftPositions and rightPositions to get them in sorted order